feat: provide better error messages in DEV

trueadm opened the pull request:
Today, Svelte 5's error message is a bit lacking. This PR aims to make the error messages much better in DEV. Previously, you'd get given the error message and some stack that likely meant nothing. With this PR, you now get the component 
stack too along with a clearer message. At also lets you know if the error occurred in different types of effect.

- We now name the effect functions, so they're clearer in the stack trace as to what they are
- We filter out Svelte internal functions from stack traces, so it makes it easier to see what is likely causing the issue
- We provide the component stack up until when the error occurred so you can what components we used in the lead up the error
- We now avoid retrying hydration or showing a mismatch for user errors. We keep the hydration mismatch warning for actual mismatches.
- Fixed a bunch of subtle things I encountered when working on this.

Rich-Harris commented on packages/svelte/src/internal/client/reactivity/effects.js:
```diff
@@ -172,7 +178,13 @@ export function user_effect(fn) {
 	validate_effect('$effect.pre');
-	return render_effect(fn);
+	var signal = render_effect(fn);
+	if (DEV) {
+		if (DEV) {
+			signal.name = '$effect.pre';
+		}
+	}
```
```suggestion
	if (DEV) {
		signal.name = '$effect.pre';
	}
```

Rich-Harris commented on packages/svelte/tests/parser-legacy/samples/action-duplicate/output.json:
```diff
@@ -31,4 +31,4 @@
 		]
 	}
-}
+}
```
can we undo all these changes so the PR is less noisy?

trueadm replied:
Sorry, fixed.

Rich-Harris commented on packages/svelte/tests/snapshot/samples/hmr/_expected/client/index.svelte.js:
```diff
@@ -11,8 +11,10 @@ function Hmr($$anchor) {
 
 if (import.meta.hot) {
 	const s = $.source(Hmr);
+	const filename = Hmr.name;
```
this isn't right. `Hmr.name` is just `'Hmr'`, we need `'Hmr.svelte'`

Rich-Harris commented:
It would be nice if we could just replace the new error stack with the original rather than including the original in the message, since stack traces are given special treatment by devtools (e.g. respecting sourcemaps, excluding things as specified by `x_google_ignoreList`). The exact treatment differs from browser to browser, but this is what I'm seeing in the playground in Firefox — it's a useless wall of text:

---

<img width="1159" alt="image" src="https://github.com/sveltejs/svelte/assets/1162160/2d96a8dd-69c6-43ec-bc20-f7f6bf888612">

---

If I do `new_error.stack = error.stack` instead of adding the stack to `component_stack_string` then I get this instead, which is much better:

---

<img width="966" alt="image" src="https://github.com/sveltejs/svelte/assets/1162160/5ace4f79-2c89-4141-9fdf-ee6e4b06ed2f">

---

Unfortunately that kills the error message in Chrome, so there's more finessing required.

Then again: if we have a working stack trace, do we need the component tree as well? What additional information does it provide? (Also, it's currently misidentifying the component responsible for the error, for me — it's blaming the component's parent.) I'm not clear to me what we gain by calling `trigger_error_boundary`.

trueadm commented:
@Rich-Harris I addressed the issues you mentioned. This PR wasn't finished yesterday when you reviewed it and it had some bugs and missing pieces still yet to do. The stack traces now properly work with sourcemaps.

Rich-Harris commented on packages/svelte/src/internal/client/runtime.js:
```diff
@@ -239,6 +242,74 @@ export function check_dirtiness(reaction) {
+		throw error;
+	}
+
+	const component_stack = [];
+
+	/** @type {import("#client").ComponentContext | null} */
+	let current_context = component_context;
+
+	while (current_context !== null) {
+		var filename = current_context.function?.filename;
+
+		if (filename) {
+			const file = filename.split('/').at(-1);
+			component_stack.push(file);
+		}
+
+		current_context = current_context.p;
+	}
+	const effect_name = effect.fn.name;
+
+	if (effect_name) {
+		// If we have an effect name, it means it happened in the template, $effect or $effect.pre.
+		Object.defineProperty(error, '__skip_hydration_retry', {
+			value: true,
+			enumerable: false
+		});
+	}
+
+	let component_stack_string =
+		`Svelte caught an error thrown by ${component_stack[0]}${effect_name ? ` during a${effect_name.startsWith('$') ? 'n' : ''} ${effect_name}` : ''}.` +
+		` You should fix this error in your code.\n\nError: ${error.message}\n\nThe component stack when this error occured:\n`;
+
+	for (const name of component_stack) {
+		component_stack_string += `\t${name}\n`;
+	}
+
+	component_stack_string += '\nThe error is located at:\n';
+
+	error.message = component_stack_string;
```
I think we can cut back on the noise — it's weird to see 'Uncaught Error: Svelte caught an error ...' for example (was it uncaught or not?). Instead of this...

<img width="1020" alt="image" src="https://github.com/sveltejs/svelte/assets/1162160/df87a077-e25f-4cc9-9209-903452cdbbf9">

...what about just this?

<img width="1020" alt="image" src="https://github.com/sveltejs/svelte/assets/1162160/d59a75cf-05c3-4eb1-8f68-eb3fc218524a">


```suggestion
	const component_stack = [];

	const effect_name = effect.fn.name;

	if (effect_name) {
		// If we have an effect name, it means it happened in the template, $effect or $effect.pre.
		Object.defineProperty(error, '__skip_hydration_retry', {
			value: true,
			enumerable: false
		});

		component_stack.push(effect_name);
	}

	/** @type {import("#client").ComponentContext | null} */
	let current_context = component_context;

	while (current_context !== null) {
		var filename = current_context.function?.filename;

		if (filename) {
			const file = filename.split('/').at(-1);
			component_stack.push(file);
		}

		current_context = current_context.p;
	}

	const indent = /Firefox/.test(navigator.userAgent) ? '  ' : '\t';
	error.message += `\n${component_stack.map((name) => `\n${indent}in ${name}`).join('')}\n`;
```

Rich-Harris commented on packages/svelte/src/internal/client/runtime.js:
```diff
@@ -260,7 +331,7 @@ export function execute_reaction_fn(signal) {
 
 	try {
-		let res = signal.fn();
+		let res = (0, signal.fn)();
```
pushed this change — it gets rid of the `[as fn]`:

<img width="361" alt="image" src="https://github.com/sveltejs/svelte/assets/1162160/4bed0ca5-8f29-48b8-90b1-65acc6feb417">

<img width="303" alt="image" src="https://github.com/sveltejs/svelte/assets/1162160/e122d3cc-f9be-4d3b-b2ec-47c7da8491c0">

trueadm replied:
Nice!

Rich-Harris commented on packages/svelte/src/internal/client/reactivity/effects.js:
```diff
@@ -249,6 +262,19 @@ export function render_effect(fn) {
+export function template_effect(fn) {
+	if (DEV) {
+		define_property(fn, 'name', {
+			value: 'Svelte template effect'
```
I think we should avoid 'effect' terminology here, as people don't necessarily think of these as effects

```suggestion
			value: '{expression}'
```

Rich-Harris commented on packages/svelte/src/internal/client/runtime.js:
```diff
@@ -239,6 +242,74 @@ export function check_dirtiness(reaction) {
+			value: true,
+			enumerable: false
+		});
+	}
```
this heuristic feels brittle. if an error occurs during component initialisation, `effect_name` is empty, and `skip_retry` will be false. AFAICT there's only one way a hydration mismatch error will be thrown, and that's `e.hydration_missing_marker_close()` in `hydration.js`. What if we just check whether the caught error had that code, rather than this `__skip_hydration_retry` stuff?

trueadm replied:
Agreed

dummdidumm replied:
Are you sure this is the only way this can happen? What if the DOM was modified by some extension and a bogus element was inserted, so the walking mechanism fails and a delete everything + remount would fix it?

trueadm replied:
> delete everything + remount would fix it?

That's not really a fix though, remounting everything is a last resort and they should really fix their problem, no?

dummdidumm replied:
What would they be able to fix if a browser extension messes with the dom?

trueadm replied:
Browser extensions break every framework, regardless of us trying to fix it during hydration. This has plagued every library/framework since the dawn of time too. I remember having issues with this and jQuery + ad blocker back in the day!

This was particularly complex with Lexical because of contentEditable. The only thing you can do is let people complain and then point them to turning off their extension. You can go further and add a MutationObserved in DEV that warns on mutations outside of Svelte, but that seems like a step too far for us to go.

Rich-Harris replied:
this seems like overkill to me (and what if the element was wrong but had the same name? we'd still be screwed)
